### ISSUE TITLE
Bug in ban mask parsing

### ISSUE THREAD

**Issue by miragu** [ircd-hybrid/ircd-hybrid#18]
I think I found a serious bug in ban parsing on server synchronization. The ban mask is incorrectly parsed when received from a remote server. It becomes just the nick part of the mask, ie, a ban on "nick!user@host" becomes simply "nick". The original server is unaffected. I'm using the 8.2.30 tarball.

The issue seems to be in add_id() when you're reusing the input of split_nuh() in a strlcpy below. The attached patch makes sure to use a dedicated buffer for it. It works for me and has been in production for 24h.

[hybrid-8.2.30-ban-fix.txt](https://github.com/ircd-hybrid/ircd-hybrid/files/4542424/hybrid-8.2.30-ban-fix.txt)

&lt;RANT&gt; I have to rant a bit. This is a very serious bug that immediately impacted my network, since most bans are in the form "&ast;!&ast;foo@host" and this gets replicated simply as "&ast;", banning everyone on the other servers and causing confusion in unaware ops. I wish some form of testing of basic functionality was implemented. &lt;RANT&gt;

**Comment by miwob** [comment]
Should be fixed now with 338c42e.

**Comment by miragu** [comment]
I also thought of that easy fix, simply using banid instead of maskptr, but then I noticed they are somewhat different even ignoring split_nuh, so I chose to isolate the split buffer and keep using maskptr in strlcpy.

That's because there's a check_string(maskptr) which sets \0 at the first whitespace. Not sure how important that is down the line but I rather not take chances. If the check does nothing, then it can be removed!

A second possible issue is that maskptr is incremented by extban_parse()'s offset. I didn't look into that code deeply, and I don't use extbans for now, but I suspect that maskptr can be different from banid if it's an extban.

Even if these issues are non-existent at the moment, IMO the chance of a future bug is high. Split_nuh is a glorified strtok and it's too dangerous to be left unchecked in C. I'll keep using my patch for now!
